cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/NotificationSubscriberService.java

```diff
@@ -313,21 +267,40 @@ protected void updateJobQueue(DatasetContext context, Notification notification)
       }
       DatasetId datasetId = DatasetId.fromString(datasetIdString);
       for (ProgramSchedule schedule : getSchedules(context, Schedulers.triggerKeyForPartition(datasetId))) {
-        addJob(new Job(schedule));
+        // long publishTimestamp = new MessageId(Bytes.fromHexString(message.getId())).getPublishTimestamp();
+        // should we use subscribe time (System.currentTimeMillis?)
+
+        JobQueueDataset jobQueue = getJobQueue(context);
+        boolean existingJobModified = false;
+        try (CloseableIterator<Job> jobs = jobQueue.getJobsForSchedule(schedule.getScheduleId())) {
+          while (jobs.hasNext()) {
+            Job job = jobs.next();
+            if (job.getState() == Job.State.PENDING_LAUNCH) {
+              // check if its PENDING_LAUNCH. If so, create a new Job to avoid the chance that the job-launching
+              // process has conflict
+              Job newJob = new SimpleJob(schedule, System.currentTimeMillis(),
+                                         Lists.newArrayList(notification), Job.State.PENDING_TRIGGER);
+              jobQueue.put(newJob);
+            } else {
+              job.getNotifications().add(notification);
```

Comment:
Should have a `Job.update(Notification notification)` to update job's notifications and change job's state accordingly